File: typescript/src/cstNode.ts
import { Ref } from './combinators.js';
import type { MatchResult } from './matchResult.js';
import { SyntaxError } from './matchResult.js';
import type { ParseResult } from './parser.js';
import { Terminal } from './terminals.js';

// -----------------------------------------------------------------------------------------------------------------

/**
 * Base class for AST and CST nodes.
 */
export abstract class Node<T extends Node<T>> {
  readonly label: string;
  readonly pos: number;
  readonly len: number;
  readonly syntaxError: SyntaxError | null;
  readonly children: readonly T[];

  protected constructor(
    label: string,
    pos: number,
    len: number,
    syntaxError: SyntaxError | null,
    children: readonly T[]
  ) {
    this.label = label;
    this.pos = pos;
    this.len = len;
    this.syntaxError = syntaxError;
    this.children = children;
  }

  getInputSpan(input: string): string {
    return input.substring(this.pos, this.pos + this.len);
  }

  toString(): string {
    return `${this.label}: pos: ${this.pos}, len: ${this.len}`;
  }

  toPrettyString(input: string): string {
    const buffer: string[] = [];
    this.buildTree(input, '', buffer, true);
    return buffer.join('');
  }

  private buildTree(input: string, prefix: string, buffer: string[], isRoot: boolean): void {
    if (!isRoot) {
      buffer.push('\n');
    }
    buffer.push(prefix);
    buffer.push(this.label);
    if (this.children.length === 0) {
      buffer.push(`: "${this.getInputSpan(input)}"`);
    }

    for (let i = 0; i < this.children.length; i++) {
      const isLast = i === this.children.length - 1;
      const childPrefix = prefix + (isRoot ? '' : isLast ? '    ' : '|   ');
      const connector = isLast ? '`---' : '|---';

      buffer.push('\n');
      buffer.push(prefix);
      buffer.push(isRoot ? '' : connector);

      this.children[i].buildTree(input, childPrefix, buffer, false);
    }
  }
}

// -----------------------------------------------------------------------------------------------------------------

/**
 * An AST node representing either a rule match or a terminal match.
 */
export class ASTNode extends Node<ASTNode> {
  private constructor(
    label: string,
    pos: number,
    len: number,
    syntaxError: SyntaxError | null,
    children: readonly ASTNode[]
  ) {
    super(label, pos, len, syntaxError, children);
  }

  static terminal(terminalMatch: MatchResult): ASTNode {
    return new ASTNode(Terminal.nodeLabel, terminalMatch.pos, terminalMatch.len, null, []);
  }

  static nonTerminal(label: string, children: readonly ASTNode[]): ASTNode {
    if (children.length === 0) {
      throw new Error('children must not be empty');
    }
    const pos = children[0].pos;
    const len = children[children.length - 1].pos + children[children.length - 1].len - pos;
    return new ASTNode(label, pos, len, null, children);
  }

  static syntaxErrorNode(se: SyntaxError): ASTNode {
    return new ASTNode(SyntaxError.nodeLabel, se.pos, se.len, se, []);
  }

  // Public factory for creating nodes in tests/CST building
  static of(label: string, pos: number, len: number, children: readonly ASTNode[]): ASTNode {
    return new ASTNode(label, pos, len, null, children);
  }
}

// -----------------------------------------------------------------------------------------------------------------

/**
 * Build an AST from a parse tree.
 */
export function buildAST(parseResult: ParseResult): ASTNode {
  const extraNode =
    parseResult.unmatchedInput !== undefined
      ? ASTNode.syntaxErrorNode(parseResult.unmatchedInput)
      : null;

  return newASTNode(parseResult.topRuleName, parseResult.root, parseResult.transparentRules, extraNode);
}

function newASTNode(
  label: string,
  refdMatchResult: MatchResult,
  transparentRules: Set<string>,
  addExtraASTNode: ASTNode | null
): ASTNode {
  const childASTNodes: ASTNode[] = [];
  collectChildASTNodes(refdMatchResult, childASTNodes, transparentRules);
  if (addExtraASTNode !== null) {
    childASTNodes.push(addExtraASTNode);
  }
  return ASTNode.nonTerminal(label, childASTNodes);
}

function collectChildASTNodes(
  matchResult: MatchResult,
  collectedAstNodes: ASTNode[],
  transparentRules: Set<string>
): void {
  if (matchResult.isMismatch) {
    return;
  }
  if (matchResult instanceof SyntaxError) {
    collectedAstNodes.push(ASTNode.syntaxErrorNode(matchResult));
  } else {
    const clause = matchResult.clause;
    if (clause instanceof Terminal) {
      collectedAstNodes.push(ASTNode.terminal(matchResult));
    } else if (clause instanceof Ref) {
      if (!transparentRules.has(clause.ruleName)) {
        collectedAstNodes.push(
          newASTNode(clause.ruleName, matchResult.subClauseMatches[0], transparentRules, null)
        );
      }
    } else {
      for (const subClauseMatch of matchResult.subClauseMatches) {
        collectChildASTNodes(subClauseMatch, collectedAstNodes, transparentRules);
      }
    }
  }
}

// -----------------------------------------------------------------------------------------------------------------

/**
 * Base class for CST nodes.
 */
export abstract class CSTNode extends Node<CSTNode> {
  protected constructor(astNode: ASTNode, children: readonly CSTNode[]) {
    super(astNode.label, astNode.pos, astNode.len, astNode.syntaxError, children);
  }
}

// -----------------------------------------------------------------------------------------------------------------

/**
 * Type alias for factory functions that create CST nodes from AST nodes.
 */
export type CSTNodeFactoryFn = (astNode: ASTNode, children: readonly CSTNode[]) => CSTNode;

// -----------------------------------------------------------------------------------------------------------------

/**
 * Build a CST from an AST.
 */
export function buildCST(
  ast: ASTNode,
  factories: Map<string, CSTNodeFactoryFn>,
  allowSyntaxErrors: boolean
): CSTNode {
  return buildCSTInternal(ast, factories, allowSyntaxErrors);
}

function buildCSTInternal(
  ast: ASTNode,
  factoriesMap: Map<string, CSTNodeFactoryFn>,
  allowSyntaxErrors: boolean
): CSTNode {
  if (ast.syntaxError !== null) {
    if (!allowSyntaxErrors) {
      throw new Error(`Syntax error: ${ast.syntaxError}`);
    }
    const errorFactory = factoriesMap.get('<SyntaxError>');
    if (errorFactory === undefined) {
      throw new Error('No factory found for <SyntaxError>');
    }
    return errorFactory(ast, []);
  }

  let factory = factoriesMap.get(ast.label);

  if (factory === undefined && ast.label === Terminal.nodeLabel) {
    factory = factoriesMap.get('<Terminal>');
  }

  if (factory === undefined) {
    throw new Error(`No factory found for rule "${ast.label}"`);
  }

  const childCSTNodes = ast.children.map((child) => buildCSTInternal(child, factoriesMap, allowSyntaxErrors));
  return factory(ast, childCSTNodes);
}

File: dart/lib/src/cst_node.dart
/// Concrete Syntax Tree (CST) node and related classes
///
/// CST nodes represent the structure of a parsed input with full syntactic detail,
/// including syntax error nodes for error recovery.
library;

/// Base class for all CST nodes.
///
/// CST nodes represent the concrete syntax structure of the input, with each node
/// corresponding to a grammar rule (non-transparent rules only) or a terminal.
abstract class CSTNode {
  /// The name of this node (rule name or `<Terminal>`)
  final String name;

  CSTNode({required this.name});

  @override
  String toString() => name;
}

/// A CST node representing a syntax error during parsing.
///
/// This node is used when the parser encounters an error and needs to represent
/// the span of invalid input. It can be included in the CST to show where errors occurred.
class CSTSyntaxErrorNode extends CSTNode {
  /// The text that caused the error
  final String text;

  /// The position in the input where the error starts
  final int pos;

  /// The length of the error span
  final int len;

  CSTSyntaxErrorNode({
    required super.name,
    required this.text,
    required this.pos,
    required this.len,
  });

  @override
  String toString() => '<SyntaxError: $name at $pos:$len>';
}

/// Metadata for creating a CST node from a parse tree node.
///
/// Each grammar rule (non-transparent) must have a corresponding factory.
/// The factory takes the rule name and actual child CST nodes,
/// and returns a CSTNode instance of type T.
///
/// Type parameter T is the specific CST node type that this factory produces.
/// It must be a subtype of CSTNode.
class CSTNodeFactory<T extends CSTNode> {
  /// The grammar rule name this factory corresponds to
  final String ruleName;

  /// Factory function that creates a CST node of type T from rule name
  /// and actual child CST nodes
  final T Function(
    String ruleName,
    List<CSTNode> children,
  ) factory;

  CSTNodeFactory({
    required this.ruleName,
    required this.factory,
  });
}

/// Exception thrown when CST factory validation fails
class CSTFactoryValidationException implements Exception {
  /// Missing rule names (rules in grammar but not in factories)
  final Set<String> missing;

  /// Extra rule names (factories provided but not in grammar)
  final Set<String> extra;

  CSTFactoryValidationException({
    required this.missing,
    required this.extra,
  });

  @override
  String toString() {
    final parts = <String>[];
    if (missing.isNotEmpty) {
      parts.add('Missing factories: ${missing.join(", ")}');
    }
    if (extra.isNotEmpty) {
      parts.add('Extra factories: ${extra.join(", ")}');
    }
    return 'CSTFactoryValidationException: ${parts.join("; ")}';
  }
}

/// Exception thrown when CST construction fails
class CSTConstructionException implements Exception {
  final String message;

  CSTConstructionException(this.message);

  @override
  String toString() => 'CSTConstructionException: $message';
}

/// Exception thrown when duplicate rule names are found in CST factories
class DuplicateRuleNameException implements Exception {
  /// The rule name that appeared more than once
  final String ruleName;

  /// The count of how many times it appeared
  final int count;

  DuplicateRuleNameException({
    required this.ruleName,
    required this.count,
  });

  @override
  String toString() =>
      'DuplicateRuleNameException: Rule "$ruleName" appears $count times in factory list';
}
